forest/tree/tree.go

```diff
@@ -0,0 +1,56 @@
+package tree
+
+import (
+	// cache "github.com/treeverse/lakefs/forest/cache_map"
+	"github.com/treeverse/lakefs/graveler"
+	"github.com/treeverse/lakefs/graveler/committed/sstable"
+)
+
+type TreePart struct {
+	PartName sstable.ID   `json:"part_name"`
+	MaxKey   graveler.Key `json:"max_path"`
+}
+type TreeSlice struct {
+	TreeSlice []TreePart
+}
+type DummyMap map[string]string // place holder for cache_map package that will be merged soon
+
+type TreesRepo struct {
+	TreesMap   DummyMap
+	PartManger sstable.Manager
+}
+
+// InitTreeRepository creates the tree cache, and stores part Manager for operations of parts (currently implemented as sstables).
+// should be called at process init.
+// decisions on who calls it and how to get a treesRepository will be taken later
+type InitTreesRepository func(manager sstable.Manager) TreeRepo
+
+type TreeRepo interface {
+	// NewTreeWriter returns a writer that uses the part manager to create a new tree
+	NewTreeWriter(splitFactor int, // average number of keys that we want to stored in a part
+		// for more detail, look at "IsSplitKey"
+		closeAsync sstable.BatchWriterCloser, // component used to close part asynchronously, and wait for all part
+		// completions when tree writing completes
+	) TreeWriter
+	// NewScannerFromTreeID accepts a tree ID, and returns an iterator over the tree
+	NewIteratorFromTreeID(treeID graveler.TreeID, start graveler.Key) (graveler.ValueIterator, error)
+	// NewIteratorFromTreeParts accept a tree in memory, returns iterator over the tree
+	NewIteratorFromTreeSlice(treeSlice TreeSlice, start graveler.Key) (graveler.ValueIterator, error)
+	// GetPartManager give components of tree package to the configured part manager.
+	// will probably change later
+	GetPartManger() sstable.Manager
+}
+
+type TreeWriter interface {
+	WriteEntry(record graveler.ValueRecord) error
```

Comment:
naming suggestions:
WriteEntry >> WriteValue

FlushIterToTree >> WriteIterator

SaveTree: split to 1. Save() which will save and return tree ID.
and 2. Merge(tree) which will create a new tree on top of an existing one and return a tree id as well.

Reply:
Good ideas. will do after merge

Reply:
Why wait?

Reply:
done